lib/deimos/active_record_consume/batch_consumption.rb

```diff
@@ -88,36 +88,85 @@ def update_database(messages)
+        import_associations(valid_upserts) unless @association_list.blank?
+      end
+
+      def save_records_to_database(record_class, key_cols, records)
+        # In-memory records contain created_at and updated_at as nil
+        # which messes up ActiveRecord-Import bulk_import method.
+        ignored_columns = %w(created_at updated_at)
+        columns = record_class.columns.map(&:name) - ignored_columns
+        # bulk_import_columns must be generated by consumer to specify which columns are key_cols
+        # as association tables do not conform to Deimos::Message keys
+        if self.respond_to?(:bulk_import_columns)
+          key_cols, columns = self.bulk_import_columns(record_class)
+        end
 
         options = if key_cols.empty?
                     {} # Can't upsert with no key, just do regular insert
-                  elsif ActiveRecord::Base.connection.adapter_name.downcase =~ /mysql/
+                  elsif my_sql_adapter?
                     {
-                      on_duplicate_key_update: :all
+                      on_duplicate_key_update: columns
                     }
                   else
                     {
                       on_duplicate_key_update: {
                         conflict_target: key_cols,
-                        columns: :all
+                        columns: columns
                       }
                     }
                   end
 
-        @klass.import!(upserts, options)
+        if self.respond_to?(:bulk_import_columns)
+           record_class.import!(columns, records, options)
+        else
+          record_class.import!(records, options)
+        end
+      end
+
+      # Imports associated objects and import them to database table
+      # The base table is expected to contain bulk_import_id column -- used for indexing associated objects with id
+      # @association_list configured on the consumer helps identify the ones required to be saved.
+      # @param entities [Array<ActiveRecord::Base>] List of ActiveRecord Models
+      # @return void
+      def import_associations(entities)
+        raise NotImplementedError unless my_sql_adapter?
+
+        validate_association_key(entities)
+        associations = @klass.reflect_on_all_associations.select { |assoc| @association_list.include?(assoc.name) }
+
+        table_by_bulk_import_id = @klass.
+          where(@bulk_import_id_column => entities.map { |e| e[@bulk_import_id_column] }).
+          select(:id, @bulk_import_id_column).
+          index_by { |e| e[@bulk_import_id_column] }
+        # update IDs in master table
+        entities.each { |entity| entity.id = table_by_bulk_import_id[entity[@bulk_import_id_column]].id }
+        associations.each do |assoc|
+          details = entities.map { |master|
+            details = Array(master.send(assoc.name))
+            details.each { |d| d.send("#{assoc.send(:foreign_key)}=", master.id) }
+            details
+          }.flatten
```

Comment:
I think I need some more comments here... I _think_ I know what this is doing but I'm having a hard time following with all the configurable values.

Reply:
I refactored this method a bit and moved that dense block in `import_assocations` method based on functionality. Please take a look when you get a chance.